lib/types/DecodedPayload.ts

```diff
@@ -1,35 +1,57 @@
+import { PluginImplementationError } from 'kuzzle';
+import { Decoder } from '../core-classes/Decoder';
 import { Measurement } from './measures/MeasureContent';
 
 /**
- * Record containing decoded measurements for each device.
- *
- * Record<deviceReference, Measurement[]>
- *
- * @example
- * const decodedPayload: DecodedPayload = {
- *   'BZH42AZF': [
- *     {
- *       deviceMeasureName: 'battery',
- *       measuredAt: 1655379939496,
- *       type: 'battery',
- *       values: { battery: 32 },
- *     },
- *   ],
- *   'IS7L8HK': [
- *     {
- *       deviceMeasureName: 'internal_temperature',
- *       measuredAt: 1655379939496,
- *       type: 'temperature',
- *       values: { temperature: -3 },
- *     },
- *     {
- *       deviceMeasureName: 'external_temperature',
- *       measuredAt: 1655379939496,
- *       type: 'temperature',
- *       values: { temperature: 39 },
- *     },
- *   ],
- * };
+ * Class containing the decoded measures.
  */
-export type DecodedPayload = Record<string, Measurement[]>;
-// @todo we should have an intermediary object
+export class DecodedPayload<TDecoder extends Decoder = Decoder> {
+  public decoder: TDecoder;
+
+  /**
+   * Measurements per device.
+   *
+   * Record<deviceReference, Measurement[]>
+   */
+  private measurementsByDevice: Record<string, Measurement[]> = {};
+
+  constructor (decoder: TDecoder) {
+    this.decoder = decoder;
+  }
+
+  get references () {
+    return Object.keys(this.measurementsByDevice);
+  }
+
+  /**
+   * Add a new measure into the decoded payload
+   *
+   * @param deviceReference Device reference
+   * @param measureName Name of the decoded measure
+   * @param measurement Measurement object
+   */
+  addMeasurement<TMeasurement extends Measurement = Measurement> (
```

Comment:
Does  generic is really useful here?

Reply:
I added it so when you want to add a measurement, the TS types will ensure that:
 - the name of the measure exists
 - the values gathered into the measures are correct

Reply:
After reading again the example, I understand. 
When I code, I (most of the time) do :

```
    measurement: BatteryMeasurement =       {
        measuredAt: Date.now(),
        type: 'battery',
        values: {
          battery: payload.batteryLevel * 100,
        };
    decodedPayload.addMeasurement(
      payload.deviceEUI,
      'mainBattery',
       measurement
      });
```

so I do not need the generic type in this case.

But in a more concise TS code style, I'm agree that GerericType is usefull in this case.